### PR TITLE
fix: update `data.property.constraints.planning` and more based on feedback

### LINKED ISSUE #31: Decide how to handle planning constraints values
Should the passport values (eg `listed`) be handled as an enum? 

Things that make this more difficult than other topics:
- Granular article 4 values that vary by council & response type (eg if `article4` is false overall, our GIS API won't return `article4.lambeth.something`) 
- In the future, if councils actually start publishing Planning Data's Article 4 "rules" dataset that will link to the geometries currently published, we'd be moving awasy from this tedious "reference" <> Planx granular variable mapping anyway

### PR DESCRIPTION
**Pull request by jessicamcinchak**
- Tightens `data.property.constraints.planning` to adhere to enum values & descriptions
  - I've made the decision here that granular, council-specific A4 variables aren't necessary to pass along as actual _variable values_ to end consumers, rather that value's name, description, and source URL will be included in the top-level `article4.entities`. Open to disagreements here, but after talking with BOPS I'm hopeful this will be a less confusing boundary between what a Planx service expects for automations and what an end consumer expects to understand which constraints have been checked
  - Resolves #31 
- Adds `data.property.address.singleLine` as this is required for feature parity in the migration guides
- Combines `data.proposal.retro.date...` &rarr; `data.proposal.date...` and makes both `start` & `completion` optional based on content clarifications from Alastair & August

For migration feature-parity testing & feedback - please doublecheck [my work here](https://opensystemslab.notion.site/opensystemslab/Proposal-for-a-Digital-Planning-Data-schema-4cc8dc9979654e459a22b0d49784c9a8#:~:text=data%20retention%20policy.-,Migration%20guide,-Planx%20currently%20generates) & let me know if any other fields are missing!

### CODE REVIEW

File: types/enums/PlanningConstraints.ts [hunk 1]
```diff
@@ -12,6 +11,10 @@ export const PlanningConstraints = {
   'designated.nationalPark.broads': 'National Park - Broads',
   'designated.SPA': 'Special Protection Area (SPA)',
   'designated.WHS': 'UNESCO World Heritage Site or buffer zone',
+  flood: 'Flood Risk Zone',
+  'flood.zone.1': 'Flood Risk Zone 1 - Low risk',
+  'flood.zone.2': 'Flood Risk Zone 2 - Medium risk',
+  'flood.zone.3': 'Flood Risk Zone 3 - High risk',
```
Comment (jessicamcinchak) [inline types/enums/PlanningConstraints.ts:17]: Optimistically including as we plan to query this dataset in near future: https://trello.com/c/a0Idx2wK/2657-add-flood-risk-zones-from-planning-data-to-list-of-datasets-queried-via-planning-constraints-check

File: types/schema/data/Property.ts [hunk 2]
```diff
@@ -96,20 +98,40 @@ type PropertyTypeMap = {
...
-interface NonOverlappingPlanningConstraint extends BasePlanningConstraint {
+type GenericPlanningConstraint<TKey extends PlanningConstraintKeys> = {
+  value: TKey;
+  description: (typeof PlanningConstraints)[TKey];
+  category?:
+    | 'Ecology'
+    | 'Flooding'
+    | 'General policy'
+    | 'Heritage and conservation'
+    | 'Trees';
```
Comment (jessicamcinchak) [inline types/schema/data/Property.ts:111]: I'm not convinced `category` is particularly meaningful to include, but BOPS is currently parsing last I heard so including this for migration feature parity

File: types/schema/data/Property.ts [hunk 3]
```diff
@@ -96,20 +98,40 @@ type PropertyTypeMap = {
...
-interface OverlappingPlanningConstraint extends BasePlanningConstraint {
+type OverlappingPlanningConstraint = {
   overlaps: true;
-  entities: string[];
-}
+  entities:
+    | {
+        name: string;
+        description?: string;
+        source?: URL;
```
Comment (jessicamcinchak) [inline types/schema/data/Property.ts:131]: Based on outcome of [this conversation](https://opendigitalplanning.slack.com/archives/C05NQJK5S9L/p1696928598458159), I imagine `entities` will grow to have an optional `rules` property or similar in the future !

**Review by DafyddLlyr — APPROVED** [review]
All looks good - the migration guide is also super thorough - I didn't spot any missing fields 👍

**Pull request merged**